numgrid/AtomGrid.cpp

```diff
@@ -0,0 +1,284 @@
+        }
+
+        int angular_off = get_angular_order(num_angular) * MAX_ANGULAR_GRID;
+
+        for (int iang = 0; iang < num_angular; iang++)
+        {
+            atom_grid_x_au.push_back(angular_x[angular_off + iang] * radial_r);
+            atom_grid_y_au.push_back(angular_y[angular_off + iang] * radial_r);
+            atom_grid_z_au.push_back(angular_z[angular_off + iang] * radial_r);
+
+            atom_grid_w.push_back(4.0 * M_PI * angular_w[angular_off + iang] *
+                                  radial_w);
+
+            num_grid_points++;
+        }
+    }
+
+    delete[] angular_x;
+    delete[] angular_y;
+    delete[] angular_z;
+    delete[] angular_w;
+}
+
+void numgrid_free_atom_grid(context_t *context)
+{
+    if (!context)
+        return;
+    delete AS_TYPE(AtomGrid, context);
+}
+AtomGrid::~AtomGrid()
+{
+    atom_grid_x_au.clear();
+    atom_grid_y_au.clear();
+    atom_grid_z_au.clear();
+    atom_grid_w.clear();
+
+    radial_atom_grid_r_au.clear();
+    radial_atom_grid_w.clear();
+}
+
+int numgrid_get_num_grid_points(const context_t *context)
+{
+    return AS_CTYPE(AtomGrid, context)->get_num_grid_points();
+}
+int AtomGrid::get_num_grid_points() const { return num_grid_points; }
+
+int numgrid_get_num_radial_grid_points(const context_t *context)
+{
+    return AS_CTYPE(AtomGrid, context)->get_num_radial_grid_points();
+}
+int AtomGrid::get_num_radial_grid_points() const
+{
+    return num_radial_grid_points;
+}
+
+void numgrid_get_grid(const context_t *context,
+                      const int num_centers,
+                      const int center_index,
+                      const double x_coordinates_au[],
+                      const double y_coordinates_au[],
+                      const double z_coordinates_au[],
+                      const int proton_charges[],
+                      double grid_x_au[],
+                      double grid_y_au[],
+                      double grid_z_au[],
+                      double grid_w[])
+{
+    return AS_CTYPE(AtomGrid, context)
+        ->get_grid(num_centers,
+                   center_index,
+                   x_coordinates_au,
+                   y_coordinates_au,
+                   z_coordinates_au,
+                   proton_charges,
+                   grid_x_au,
+                   grid_y_au,
+                   grid_z_au,
+                   grid_w);
+}
+void AtomGrid::get_grid(const int num_centers,
+                        const int center_index,
+                        const double x_coordinates_au[],
+                        const double y_coordinates_au[],
+                        const double z_coordinates_au[],
+                        const int proton_charges[],
+                        double grid_x_au[],
+                        double grid_y_au[],
+                        double grid_z_au[],
+                        double grid_w[]) const
+{
+    for (int ipoint = 0; ipoint < num_grid_points; ipoint++)
+    {
+        grid_x_au[ipoint] =
```

Comment:
These could possibly be implemented with calls to `std::transform` without looping explicitly:
```
std::transform(atom_grid_x_au.cbegin(), atom_grid_x_au.cend(),
                      &x_coordinates_au[0], &grid_x_au[0], std::plus<double>());
```
I haven't tested this explicitly and it might need to be modified if you don't want to user C++11

Reply:
I will try this.